src/tree.ml

```diff
@@ -0,0 +1,724 @@
+
+(** {1 Types for decision trees}
+
+    The types involved in the definition of decision trees are given in module
+    {!module:Tree_types} (they could not be defined here as this would lead to
+    a cyclic dependency).
+
+    {b Example:} let us consider the rewrite system for symbol [f] defined as:
+    - [f Z     (S m) → S m],
+    - [f n     Z     → n] and
+    - [f (S n) (S m) → S (S m)].
+
+    A possible decision tree might be
+    {v
+    ├─?─∘─Z─∘     → n
+    ├─Z─∘─Z─∘     → n
+    │   └─S─∘─?─∘ → S m
+    ├─S─∘─Z─∘     → n
+    └─S─∘─?─∘     → S (S m)
+    v}
+    with [∘] being a node (with an omitted label) and [─u─] being an edge with
+    a matching on symbol [u] or a variable or wildcard when [?]. Typically the
+    portion [S─∘─Z] is made possible by a swap. *)
+
+(** Representation of a rewriting rule RHS (or action) as given in the type of
+    rewriting rules (see {!field:Terms.rhs}). In decision trees, we will store
+    a RHS in every leaf since they correspond to matched rules. *)
+type rhs = (term_env, term) Bindlib.mbinder
+
+(** Representation of a branching condition (see {!type:Terms.tree_cond}). *)
+type tree_cond = term Tree_types.tree_cond
+
+(** Representation of a tree (see {!type:Terms.tree}). *)
+type tree = (term, rhs) Tree_types.tree
+
+(** {1 Conditions for decision trees}
+
+    The decision trees used for pattern matching include binary nodes carrying
+    conditions (see constructor {!constructor:Cond} of {!type:Tree_types.tree}
+    for more details). These conditions are tested during evaluation to select
+    which of the two subsequent branches to follow.
+
+    There are two forms of conditions:
+    - convertibility conditions (see {!constructor:Tree_types.CondNL}),
+    - free variable conditions (see {!constructor:Tree_types.CondFV}).
+
+    Convertibility conditions are used we have non left-linear rewriting rules
+    such as [f &x &x (s &y) → r]. In this case we need to test whether the two
+    terms at position [{0}] and [{1}] (corresponding to [&x]) are convertible:
+    the rule can only be apply if that is the case. Note that in general there
+    may be more than two occurrences of a variables. In that case we will need
+    to check convertibility between more than two terms.
+
+    Free variable constraints are used to verify which variables are free in a
+    term. If there is a rule of the form [f (λ x y, &Y[y]) → &Y], then we need
+    to check that the term at position [{0.0}] depends only on [y]. *)
+
+(** Module providing a representation for pools of conditions. *)
+module CP = struct
+  (** Functional sets of pairs of integers. *)
+  module PSet = Set.Make(
+    struct
+      type t = int * int
+
+      let compare : t -> t -> int = fun (i1,i2) (j1,j2) ->
+        match i1 - j1 with 0 -> i2 - j2 | k -> k
+    end)
+
+  (** A pool of (convertibility and free variable) conditions. *)
+  type t =
+    { nl_partial : int IntMap.t
```

Comment:
Could you explain what exactly is the idea with this map? I'm not completely sure I follow, and the comment is not very helpful to get an intuition (and looking at the code I'm not sure why this should be correct).

Reply:
Since the constraint involve two variables and we inspect one variable at a time, we must memorise which variables have been seen.  This mapping allows to remember which variable pointed to which slot in the environment.  If the variable inspected points to the same slot as another one in this mapping, then it means it must be convertible with that previous variable